comm/handlers.go

```diff
@@ -14,65 +18,296 @@ import (
+
+	defer func() {
+		err := recover()
+		if err != nil {
+			HandleError(response, err)
+		}
+	}()
+
+	account, err := ValidateRequest(response, request)
+	if err != nil {
+		HandleError(response, err)
+		return
+	}
+
+	cmd := GetClientRequest(request)
+
+	command := jobs.NewCommand(cmd)
+
+	command = Global().Sched.AddScheduledCommand(account.Username, command)
+	fmt.Println("Scheduling Work", command.JobId)
+
+	WriteResponse(response, "JobId: ", command.JobId)
+}
+
+func WriteResponse(response http.ResponseWriter, args ...interface{}) {
+	buffer := fmt.Sprint(args...)
+	response.Write([]byte(buffer))
+	response.Write([]byte("\n"))
+	response.(http.Flusher).Flush()
+}
+
+// Execute a command right away
+func Remove(response http.ResponseWriter, request *http.Request) {
+
+	defer func() {
+		err := recover()
+		if err != nil {
+			HandleError(response, err)
+		}
+	}()
+
+	account, err := ValidateRequest(response, request)
+	if err != nil {
+		HandleError(response, err)
+		return
+	}
+
+	cmd := GetClientRequest(request)
+
+	status := Global().Sched.RemoveUserCommand(account.Username, cmd.JobId)
+
+	WriteResponse(response, "Status: ", status)
+}
+
+// Execute a command right away
+func Tail(response http.ResponseWriter, request *http.Request) {
+
+	defer func() {
+		err := recover()
+		if err != nil {
+			HandleError(response, err)
+		}
+	}()
+
+	account, err := ValidateRequest(response, request)
+	if err != nil {
+		HandleError(response, err)
+		return
+	}
+
+	cmd := GetClientRequest(request)
+
+	entry := Global().Active.FindJobStatus(account.Username, cmd.JobId)
+
+	StreamResponse(entry, response)
+}
+
+// Execute a command right away
+func Output(response http.ResponseWriter, request *http.Request) {
+
+	defer func() {
+		err := recover()
+		if err != nil {
+			HandleError(response, err)
+		}
+	}()
+
+	account, err := ValidateRequest(response, request)
+	if err != nil {
+		HandleError(response, err)
+		return
+	}
+
+	cmd := GetClientRequest(request)
+
+	entry := Global().Active.FindJobStatus(account.Username, cmd.JobId)
+
+	if entry != nil {
+		StreamResponse(entry, response)
+	} else {
+		WriteResponse(response, "Unknown Job Id: ", cmd.JobId)
+	}
+}
+
+// Execute a command right away
+func Status(response http.ResponseWriter, request *http.Request) {
+	defer func() {
+		err := recover()
+		if err != nil {
+			HandleError(response, err)
+		}
+	}()
+
+	account, err := ValidateRequest(response, request)
+	if err != nil {
+		HandleError(response, err)
+		return
+	}
+
+	cmd := GetClientRequest(request)
+
+	entry := Global().Active.FindJobStatus(account.Username, cmd.JobId)
+	if entry != nil {
+		if entry.IsRunning {
+			WriteResponse(response, "Status: ", "Running")
+		} else {
+			WriteResponse(response, "Status: ", "Finished")
+		}
+	} else {
+		WriteResponse(response, "Status: ", "Pending")
+	}
+}
+
+// Handle logging and sending back and error
+func HandleError(response http.ResponseWriter, err interface{}) {
```

Comment:
It's because I'm using it in recover() as well, which seemed to return its param as an interface{}.

Reply:
Hm, to be honest, this is a bit out of my ballpark because I rarely use `recover`. My understanding is that `recover` returns whatever is given to `panic`, hence the empty interface. But this doesn't necessarily need to be an error.

Particularly, I think that this `HandleError` method will panic if the recovered value itself is a string. As an example: https://play.golang.org/p/37jhPYnN-z_X

I would prefer to change `HandleError` to accept an `error` type instead to make this safe and have the function that `recover`s from a panic deal with casting the value into an acceptable error type.

Reply:
Yea, it's pretty hacky. If I were working normally I would have tried to wire in something above all of this to deal with panics, validations, etc. It's very kludgy and repetitive right now.